topi/python/topi/nn/convolution.py

```diff
@@ -1,9 +1,232 @@
+from .. import target as _target
+
+# workload description of convolution
+Workload = namedtuple('Workload',
+                      ['height', 'width', 'in_filter', 'out_filter',
+                       'hkernel', 'wkernel', 'hpad', 'wpad', 'hstride', 'wstride'])
+
+# schedule description of spatial
+SpatialPack = namedtuple('SpatialPack',
+                         ['vh', 'vw', 'vc', 'ba', 'bc', 'unroll'])
+
+# schedule description of im2col
+Im2ColPack = namedtuple('Im2ColPack',
+                        ['vp', 'vq', 'ba', 'bc', 'unroll'])
+
+# workloads of resnet18 on imagenet
+_WORKLOADS = [
+    Workload(224, 224, 3, 64, 7, 7, 3, 3, 2, 2),
+    Workload(56, 56, 64, 64, 3, 3, 1, 1, 1, 1),
+    Workload(56, 56, 64, 64, 1, 1, 0, 0, 1, 1),
+    Workload(56, 56, 64, 128, 3, 3, 1, 1, 2, 2),
+    Workload(56, 56, 64, 128, 1, 1, 0, 0, 2, 2),
+    Workload(28, 28, 128, 128, 3, 3, 1, 1, 1, 1),
+    Workload(28, 28, 128, 256, 3, 3, 1, 1, 2, 2),
+    Workload(28, 28, 128, 256, 1, 1, 0, 0, 2, 2),
+    Workload(14, 14, 256, 256, 3, 3, 1, 1, 1, 1),
+    Workload(14, 14, 256, 512, 3, 3, 1, 1, 2, 2),
+    Workload(14, 14, 256, 512, 1, 1, 0, 0, 2, 2),
+    Workload(7, 7, 512, 512, 3, 3, 1, 1, 1, 1),
+]
+
+# platform specific schedule
+_CONV_SCHEDULE = {}
+
+# platform specific declaration
+_CONV_DECLARATION = {}
+
+def convolution(data, kernel, stride, padding, layout='NCHW'):
+    """Convolution operator.
+
+    Parameters
+    ----------
+    input : tvm.Tensor
+        4-D with shape [batch, in_channel, in_height, in_width]
+
+    filter : tvm.Tensor
+        4-D with shape [num_filter, in_channel, filter_height, filter_width]
+
+    stride : int or a list/tuple of two ints
+        stride size, or [stride_height, stride_width]
+
+    padding : int or a list/tuple of two ints
+        padding size, or [pad_height, pad_width]
+
+    layout : str
+        layout of data
+
+    Returns
+    -------
+    output : tvm.Tensor
+        4-D with shape [batch, out_channel, out_height, out_width]
+    """
+    # search platform specific declaration first
+    target = _target.current_target()
+    if target in _CONV_DECLARATION:
+        return _CONV_DECLARATION[target](data, kernel, stride, padding, layout)
+
+    # default declaration
+    if layout == 'NCHW':
+        conv2d_nchw(data, kernel, stride, padding)
+    elif layout == 'HWCN':
+        conv2d_hwcn(data, kernel, stride, padding)
+    else:
+        raise ValueError("not support this layout {} yet".format(layout))
+
+
+def _get_workload(data, kernel, stride, padding):
+    """ Get the workload structure. """
+    _, CI, IH, IW = [x.value for x in data.shape]
+    CO, _, KH, KW = [x.value for x in kernel.shape]
+    HPAD, WPAD, _, _ = get_pad_tuple(padding, kernel)
+    if isinstance(stride, (tuple, list)):
+        HSTR, WSTR = stride
+    else:
+        HSTR, WSTR = stride, stride
+    return Workload(IH, IW, CI, CO, KH, KW, HPAD, WPAD, HSTR, WSTR)
+
+def _get_schedule(wkl, target=None):
```

Comment:
Seems this is the only place _CONV_SCHEDULE get referenced, can simply move to the target specific place